### PR TITLE
Add additional callbacks for ConnectionEvents

### PR DESCRIPTION
**Pull request by somasun**
- Added onStreamsResumed() and onStreamsPaused() callbacks
- Added ConnectionEvents tests

### CODE REVIEW

File: rsocket/RSocketConnectionEvents.h [hunk 1]
```diff
@@ -8,12 +8,35 @@ class exception_wrapper;
...
+// The application should implement this interface to get called-back
+// on network events.
 class RSocketConnectionEvents {
  public:
   virtual ~RSocketConnectionEvents() = default;
 
+  // This method gets called when the underlying transport is connected to the
+  // remote side.  This does not necessarily mean that the RSocket connection
+  // will be successful.  As an example, the transport might get reconnected
+  // for an existing RSocketStateMachine.  But the resumption at RSocket layer
```
Comment (alexmalyshev) [inline rsocket/RSocketConnectionEvents.h]: Nit: `But the resumption at RSocket layer` --> `But resumption at the RSocket layer`

File: test/ConnectionEventsTest.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  MOCK_METHOD1(onDisconnected, void(const folly::exception_wrapper&));
+  MOCK_METHOD1(onStreamsPaused, void(const folly::exception_wrapper&));
+  MOCK_METHOD0(onStreamsResumed, void());
+  MOCK_METHOD1(onClosed, void(const folly::exception_wrapper&));
+};
+
+} // anonymous namespace
+
+TEST(ConnectionEventsTest, SimpleStream) {
+  auto serverConnEvents = std::make_shared<HelloServiceConnectionEvents>();
```
Comment (alexmalyshev) [inline test/ConnectionEventsTest.cpp]: Could these be `StrictMock`?

File: rsocket/RSocketConnectionEvents.h [hunk 3]
```diff
@@ -8,12 +8,35 @@ class exception_wrapper;
...
   virtual void onDisconnected(const folly::exception_wrapper&) {}
+
+  // This gets called when the RSocketStateMachine is closed.  You cant use this
+  // RSocketStateMachine anymore.
   virtual void onClosed(const folly::exception_wrapper&) {}
+
+  // This gets called when no more frames can be sent over the RSocket streams.
+  // This typically happens immediately after onDisconnected(). The streams can
+  // be resumed after onStreamsResumed() event.
+  virtual void onStreamsPaused(const folly::exception_wrapper&) {}
```
Comment (lehecka) [inline rsocket/RSocketConnectionEvents.h]: I think the parameter is not necessary. See the other comment too.

File: rsocket/statemachine/RSocketStateMachine.cpp [hunk 4]
```diff
@@ -160,7 +160,11 @@ void RSocketStateMachine::disconnect(folly::exception_wrapper ex) {
   }
 
   closeFrameTransport(std::move(ex), StreamCompletionSignal::CONNECTION_END);
-  pauseStreams();
+
+  if (connectionEvents_) {
+    connectionEvents_->onStreamsPaused(ex);
```
Comment (lehecka) [inline rsocket/statemachine/RSocketStateMachine.cpp]: parameter ex is already moved on line 162 so you are never sending an exception. I would suggest to remove the param.

**Pull request merged**